File: cpp/lib/connection/Driver.cpp
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "typedb/connection/driver.hpp"

#include "inc/macros.hpp"

using namespace TypeDB;

namespace TypeDB {

Credential::Credential(const std::string& username, const std::string& password, bool withTLS, const std::string& customRootCAStr) {
    const char* customRootCA = customRootCAStr != "" ? customRootCAStr.c_str() : nullptr;
    auto p = _native::credential_new(username.c_str(), password.c_str(), customRootCA, withTLS);
    TypeDBDriverException::check_and_throw();
    credentialNative = std::move(NativePointer<_native::Credential>(p, _native::credential_drop));
}

_native::Credential* Credential::getNative() const {
    return credentialNative.get();
}


_native::Connection* connectToCore(const std::string& coreAddress) {
    auto p = _native::connection_open_core(coreAddress.c_str());
    TypeDBDriverException::check_and_throw();
    return p;
}

_native::Connection* connectToEnterprise(const std::vector<std::string>& enterpriseAddresses, _native::Credential* nativeCredential) {
    std::vector<const char*> addressesNative;
    for (auto& addr : enterpriseAddresses)
        addressesNative.push_back(addr.c_str());
    addressesNative.push_back(nullptr);
    auto p = _native::connection_open_enterprise(addressesNative.data(), nativeCredential);
    TypeDBDriverException::check_and_throw();
    return p;
}

Driver::Driver()
    : Driver(nullptr) {}

Driver::Driver(const std::string& coreAddress)
    : Driver(connectToCore(coreAddress)) {
    TypeDBDriverException::check_and_throw();
}

Driver::Driver(const std::vector<std::string>& enterpriseAddresses, const Credential& credential)
    : Driver(connectToEnterprise(enterpriseAddresses, credential.getNative())) {
    TypeDBDriverException::check_and_throw();
}

Driver::Driver(_native::Connection* conn) noexcept
    : connectionNative(conn, _native::connection_close),
      databases(this->connectionNative.get()),
      users(this->connectionNative.get()) {}

Driver::Driver(Driver&& from)
    : connectionNative(std::move(from.connectionNative)),
      databases(std::move(from.databases)),
      users(std::move(from.users)) {}

Driver& Driver::operator=(Driver&& from) {
    connectionNative = std::move(from.connectionNative);
    from.connectionNative = nullptr;
    databases = std::move(from.databases);
    users = std::move(from.users);
    return *this;
}

bool Driver::isOpen() {
    return connectionNative != nullptr && _native::connection_is_open(connectionNative.get());
}

void Driver::close() {
    connectionNative.reset();
}

Session Driver::session(const std::string& database, SessionType sessionType, const Options& options) {
    CHECK_NATIVE(connectionNative);
    return databases.session(database, sessionType, options);
}

User Driver::user() {
    CHECK_NATIVE(connectionNative);
    return users.getCurrentUser();
}

}  // namespace TypeDB

File: csharp/Connection/Driver.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;

using TypeDB.Driver;
using TypeDB.Driver.Api;
using TypeDB.Driver.Common;
using TypeDB.Driver.Common.Validation;
using TypeDB.Driver.Connection;
using TypeDB.Driver.User;

namespace TypeDB.Driver.Connection
{
    public class TypeDBDriver : NativeObjectWrapper<Pinvoke.TypeDBDriver>, IDriver
    {
        private readonly IDatabaseManager _databaseManager;
        private readonly UserManager _userManager;

        public TypeDBDriver(string address, Credentials credentials, DriverOptions driverOptions)
            : base(Open(address, credentials, driverOptions))
        {
            _databaseManager = new DatabaseManager(NativeObject);
            _userManager = new UserManager(NativeObject);
        }

        public TypeDBDriver(ISet<string> addresses, Credentials credentials, DriverOptions driverOptions)
            : base(Open(addresses, credentials, driverOptions))
        {
            _databaseManager = new DatabaseManager(NativeObject);
            _userManager = new UserManager(NativeObject);
        }

        public TypeDBDriver(IDictionary<string, string> addressTranslation, Credentials credentials, DriverOptions driverOptions)
            : base(Open(addressTranslation, credentials, driverOptions))
        {
            _databaseManager = new DatabaseManager(NativeObject);
            _userManager = new UserManager(NativeObject);
        }

        private static Pinvoke.TypeDBDriver Open(string address, Credentials credentials, DriverOptions driverOptions)
        {
            Validator.RequireNonNull(address, nameof(address));
            Validator.RequireNonNull(credentials, nameof(credentials));
            Validator.RequireNonNull(driverOptions, nameof(driverOptions));
            try
            {
                return Pinvoke.typedb_driver.driver_new_with_description(
                    address,
                    credentials.NativeObject,
                    driverOptions.NativeObject,
                    IDriver.Language);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        private static Pinvoke.TypeDBDriver Open(ISet<string> addresses, Credentials credentials, DriverOptions driverOptions)
        {
            Validator.RequireNonNull(addresses, nameof(addresses));
            Validator.RequireNonNull(credentials, nameof(credentials));
            Validator.RequireNonNull(driverOptions, nameof(driverOptions));
            try
            {
                return Pinvoke.typedb_driver.driver_new_with_addresses_with_description(
                    addresses.ToArray(),
                    credentials.NativeObject,
                    driverOptions.NativeObject,
                    IDriver.Language);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        private static Pinvoke.TypeDBDriver Open(IDictionary<string, string> addressTranslation, Credentials credentials, DriverOptions driverOptions)
        {
            Validator.RequireNonNull(addressTranslation, nameof(addressTranslation));
            Validator.RequireNonNull(credentials, nameof(credentials));
            Validator.RequireNonNull(driverOptions, nameof(driverOptions));
            try
            {
                string[] publicAddresses = addressTranslation.Keys.ToArray();
                string[] privateAddresses = addressTranslation.Values.ToArray();
                return Pinvoke.typedb_driver.driver_new_with_address_translation_with_description(
                    publicAddresses,
                    privateAddresses,
                    credentials.NativeObject,
                    driverOptions.NativeObject,
                    IDriver.Language);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public bool IsOpen()
        {
            if (!NativeObject.IsOwned())
            {
                return false;
            }
            return Pinvoke.typedb_driver.driver_is_open(NativeObject);
        }

        public IDatabaseManager Databases
        {
            get { return _databaseManager; }
        }

        public IUserManager Users
        {
            get { return _userManager; }
        }

        public ITransaction Transaction(string database, TransactionType type)
        {
            return Transaction(database, type, new TransactionOptions());
        }

        public ITransaction Transaction(string database, TransactionType type, TransactionOptions options)
        {
            Validator.RequireNonNull(database, nameof(database));
            Validator.RequireNonNull(options, nameof(options));
            try
            {
                Pinvoke.Transaction nativeTransaction = Pinvoke.typedb_driver.transaction_new(
                    NativeObject,
                    database,
                    (Pinvoke.TransactionType)type,
                    options.NativeObject);

                return new Transaction(this, nativeTransaction, type, options);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public ServerVersion GetServerVersion(ServerRouting? serverRouting = null)
        {
            try
            {
                return new ServerVersion(
                    Pinvoke.typedb_driver.driver_server_version(
                        NativeObject,
                        ServerRouting.GetNativeValue(serverRouting)));
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public ISet<IServer> GetServers(ServerRouting? serverRouting = null)
        {
            try
            {
                return new NativeEnumerable<Pinvoke.Server>(
                    Pinvoke.typedb_driver.driver_servers(
                        NativeObject,
                        ServerRouting.GetNativeValue(serverRouting)))
                    .Select(s => (IServer)new ServerImpl(s))
                    .ToHashSet();
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public IServer? GetPrimaryServer(ServerRouting? serverRouting = null)
        {
            try
            {
                Pinvoke.Server? nativeServer = Pinvoke.typedb_driver.driver_primary_server(
                    NativeObject,
                    ServerRouting.GetNativeValue(serverRouting));
                if (nativeServer != null)
                {
                    return new ServerImpl(nativeServer);
                }
                return null;
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void RegisterServer(long serverId, string address)
        {
            try
            {
                Pinvoke.typedb_driver.driver_register_server(NativeObject, serverId, address);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void DeregisterServer(long serverId)
        {
            try
            {
                Pinvoke.typedb_driver.driver_deregister_server(NativeObject, serverId);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void UpdateAddressTranslation(IDictionary<string, string> addressTranslation)
        {
            Validator.RequireNonNull(addressTranslation, nameof(addressTranslation));
            try
            {
                string[] publicAddresses = addressTranslation.Keys.ToArray();
                string[] privateAddresses = addressTranslation.Values.ToArray();
                Pinvoke.typedb_driver.driver_update_address_translation(
                    NativeObject, publicAddresses, privateAddresses);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void Close()
        {
            if (!NativeObject.IsOwned())
            {
                return;
            }

            try
            {
                Pinvoke.typedb_driver.driver_force_close(NativeObject);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void Dispose()
        {
            Close();
        }
    }
}
